modoboa/lib/email_utils.py

```diff
@@ -166,7 +169,24 @@ def _parse_text(self, msg, level=0):
 
         if content_type in ["plain", "html"]:
             encoding = msg.get_content_charset() or "utf-8"
-            mail_text = smart_text(msg.get_payload(decode=True), encoding=encoding, errors="replace")
+            mail_text_bytes = msg.get_payload(decode=True)
+            try:
+                mail_text = mail_text_bytes.decode(encoding)
+            except UnicodeDecodeError:
+                encoding = chardet.detect(mail_text_bytes)
+                try:
+                    mail_text = mail_text_bytes.decode(
+                        encoding["encoding"], "replace"
+                    )
+                except (TypeError, UnicodeDecodeError) as exc:
+                    six.raise_from(
+                        InternalError(
+                            _(
+                                "unable to determine encoding of e-mail message"
+                            )
+                        ),
+                        exc
+                    )
```

Comment:
This code looks very similar to the one you put in the quarantine PR. There is also a ``decode()`` function is th same file. Maybe we could refactor them a bit?

Reply:
good point, refactored